### ISSUE TITLE
Get-AuthenticodeSignature needs to use embedded signatures

### ISSUE THREAD

**Issue by pronichkin** [PowerShell/PowerShell#23820]
### Summary of the new feature / enhancement

This is continuation of #8401 but with some new context.

TL;DR we need `Get-AuthenticodeSignature` to consider embedded (Authenticode) signatures even if the file is catalog signed.

Long explanation

Windows ships with the boot loader signed by “Microsoft Windows Production PCA 2011” certificate. This certificate has [known security issues](https://www.microsoft.com/en-us/security/blog/2023/04/11/guidance-for-investigating-attacks-using-cve-2022-21894-the-blacklotus-campaign/), and will eventually be [revoked from UEFI](https://support.microsoft.com/help/5025885), rendering the old boot loader unusable. Therefore, Windows updates distribute a new boot loader signed by “Windows UEFI CA 2023” certificate. This boot loader is meant to eventually replace the old boot loader on all machines.

Because there are a lot of moving parts in the process (OS updates, UEFI trust list, OEMs, 3rd party boot loaders and UEFI modules, etc.) the process of replacing the boot loader is convoluted and spans across multiple phases. Not everyone gets the new boot loader at once, and it's not always safe to have it replaced on your machine right away. Extreme caution is advised.

With that, a user has a legitimate need to know how this or that boot loader are signed. This is easy to illustrate since both old and new boot loaders are present on a sufficiently up to date Windows machine.

* the original boot loader is located at `C:\Windows\Boot\EFI\bootmgfw.efi`
* the updated boot loader is located at `C:\Windows\Boot\EFI_EX\bootmgfw_EX.efi`

(However, the problem is not limited to just these two particular files, since other boot loaders may be involved.)

The easiest way to check the signature is GUI. File properties in Windows Explorer do reflect the differences in signatures correctly, and [it is currently documented](https://support.microsoft.com/help/5037167) as the only way a user can validate those signatures. (Scroll down the page and expand _“Step 2: Install the PCA2023-signed boot manager”_ section to see that guidance.)

![image](https://github.com/PowerShell/PowerShell/assets/17203954/aedacb90-a45f-45cc-bb05-ebf0d637d077)
![image](https://github.com/PowerShell/PowerShell/assets/17203954/e31fe302-6074-4013-832d-5692db0bbcb4)

There are a lot of obvious reasons why GUI is not a great choice for this task. For instance, one cannot use Windows Explorer to peek into EFI system partition (ESP). Therefore, official documentation includes a multi-step process which includes assigning a temporary drive letter to that partition and using command line tools to copy the file over, and only after that using Explorer to observe the properties of the copy. This is highly inefficient, not scalable and error prone. Ideally, it should be one or two lines in PowerShell to validate the **actual** boot loader in use.

Unfortunately, the Windows boot loader files are dual-signed, and the catalog signature is not updated in the new boot loader. (Perhaps because UEFI only cares for the embedded signature.) Therefore, `Get-AuthenticodeSignature` is not very helpful here.

```
PS C:\Users\artemp> $old = Get-Item -Path 'C:\Windows\Boot\EFI\bootmgr.efi'
PS C:\Users\artemp> $new = Get-Item -Path 'C:\Windows\Boot\EFI_EX\bootmgfw_EX.efi'
PS C:\Users\artemp> Get-AuthenticodeSignature -FilePath $old.FullName | Select-Object -ExpandProperty 'SignerCertificate' | Select-Object -ExpandProperty 'issuer'
CN=Microsoft Windows Production PCA 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
PS C:\Users\artemp> Get-AuthenticodeSignature -FilePath $new.FullName | Select-Object -ExpandProperty 'SignerCertificate' | Select-Object -ExpandProperty 'issuer'
CN=Microsoft Windows Production PCA 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
PS C:\Users\artemp>
```

### Proposed technical implementation details (optional)

One option is to use `sigcheck.exe` from Sysinternals. It does show the difference in authenticode signatures as expected. But it's not very suitable for automation.
```
PS C:\Users\artemp> sigcheck.exe -noBanner -i $old.FullName | Select-Object -First 10
c:\windows\boot\efi\bootmgr.efi:
        Verified:       Signed
        Link date:      11:21 PM 10/17/1964
        Signing date:   12:50 AM 3/22/2024
        Catalog:        c:\windows\boot\efi\bootmgr.efi
        Signers:
           Microsoft Windows
                Cert Status:    Valid
                Valid Usage:    NT5 Crypto, Code Signing
                Cert Issuer:    Microsoft Windows Production PCA 2011
PS C:\Users\artemp> sigcheck.exe -noBanner -i $new.FullName | Select-Object -First 10
c:\windows\boot\efi_ex\bootmgfw_EX.efi:
        Verified:       Signed
        Link date:      11:14 PM 12/10/1983
        Signing date:   9:55 AM 3/29/2024
        Catalog:        c:\windows\boot\efi_ex\bootmgfw_EX.efi
        Signers:
           Microsoft Windows
                Cert Status:    Valid
                Valid Usage:    Code Signing, NT5 Crypto
                Cert Issuer:    Windows UEFI CA 2023
PS C:\Users\artemp>
```

We can also rely on workaround suggested in #8401. Currently it looks like a best option, but it's not very obvious.
```
PS C:\Users\artemp> $certificate = [System.Security.Cryptography.X509Certificates.X509Certificate]::CreateFromSignedFile( $old.FullName )
PS C:\Users\artemp> [System.Security.Cryptography.X509Certificates.X509Certificate2]::new( $certificate ) | Select-Object -ExpandProperty 'issuer'
CN=Microsoft Windows Production PCA 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
PS C:\Users\artemp> $certificate = [System.Security.Cryptography.X509Certificates.X509Certificate]::CreateFromSignedFile( $new.FullName )
PS C:\Users\artemp> [System.Security.Cryptography.X509Certificates.X509Certificate2]::new( $certificate ) | Select-Object -ExpandProperty 'issuer'
CN=Windows UEFI CA 2023, O=Microsoft Corporation, C=US
PS C:\Users\artemp>
```

Finally, it looks like the community has recognized the shortcoming of `Get-AuthenticodeSignature` because alternative solutions emerge, such as [Get-DigitalSignature](https://www.powershellgallery.com/packages/PSReflect-Functions/2.0.0/Content/Examples%5CGet-DigitalSignature.ps1) script. Unfortunately, it has a large dependency which is difficult to audit and therefore it's not very suitable for use in security-focused production environments.

**Comment by jborean93** [comment]
Just an FYI be careful with using `CreateFromSignedFile` as it does a very naive scan the raw bytes for a PKCS7 byte sequence rather than actually extracting the signature based on the file extension.

While `Get-AuthenticodeSignature` does have a `-Content` and `-SourcePathOrExtension` parameter to rovide the raw bytes of the file manually it is [currently hardcoded to use the PowerShell SIP provider](https://github.com/PowerShell/PowerShell/blob/da1ca4a7266e2c9e43f96a38aeea8092ace71cca/src/System.Management.Automation/security/Authenticode.cs#L480-L488) so this won't work for exe/efi files :(

Adding in a switch parameter to bypass the API that checks the catalog files should be a pretty easy thing to implement https://github.com/PowerShell/PowerShell/blob/da1ca4a7266e2c9e43f96a38aeea8092ace71cca/src/System.Management.Automation/security/Authenticode.cs#L282-L286.

**Comment by jborean93** [comment]
I need to write some tests for CI but I've tested https://github.com/PowerShell/PowerShell/pull/23821 with the example file in https://github.com/PowerShell/PowerShell/issues/8401 and it returns both the catalogue cert by default and the embedded cert with `-EmbeddedSignature` set

![image](https://github.com/PowerShell/PowerShell/assets/8462645/d1022243-56f6-4309-acf0-9f534c4c9c54)

My efi file in your example still seems to be the same signature so I cannot test that example myself.

**Comment by iSazonov** [comment]
We could add new test C# project, sign the exe and use in the test.

**Comment by iSazonov** [comment]
Or reuse test\tools\TestExe

**Comment by jborean93** [comment]
It should be possible to just use a new ps1 file. Even generating an exe at runtime isn’t that hard to do. The hard bit is figuring out how to setup the catalog file so it is used in a normal check. I’ve found some APIs I just need to try them out.

**Comment by tylermontneyacc** [comment]
@pronichkin Thanks for reporting this, saved me a considerable amount of frustration.

**Comment by Rickth64** [comment]
Thanks. +1

**Comment by vyv03354** [comment]
Another workaround: Use `Get-PfxCertificate`.
```
PS C:\WINDOWS\system32> (Get-PfxCertificate C:\Windows\Boot\EFI\bootmgfw.efi).Issuer
CN=Microsoft Windows Production PCA 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US
PS C:\WINDOWS\system32> (Get-PfxCertificate C:\Windows\Boot\EFI_EX\bootmgfw_EX.efi).Issuer
CN=Windows UEFI CA 2023, O=Microsoft Corporation, C=US
PS C:\WINDOWS\system32>
```
